Add Sapho Chart

lodotek opened the pull request.

Commit:
Initial Commit

prydonius commented:
@k8s-bot ok to test

prydonius commented:
Thanks for the submission @lodotek!

Can you remove the vendored mysql in the `charts/` directory? You should be able to use `helm dep update` to pull in the packaged mysql chart which will be ignored in git. The CI will automatically do this when packaging the chart for the repository.

Also, AFAICT this meets the [requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements) to be included in the stable repo, so you can move this to the stable folder :).

prydonius commented on incubator/sapho/values.yaml:
```diff
@@ -0,0 +1,37 @@
+saphoDBport: 3306
+saphoDBtype: mysql
+saphoDBuser: root
+saphoDBpass: 7iro7Ux6Xl
```
looks like this isn't being used

Commit:
cleaned up some unused values...

Commit:
Delete values.yaml

Commit:
cleanup

lodotek commented:
@prydonius I have made the requested changes, please test again. 

Thank you!

aronchick commented:
@prydonius I'd love to demo this soon - any chance this is ready to go?

viglesiasce commented on incubator/sapho/values.yaml:
```diff
@@ -0,0 +1,34 @@
+    cpu: 300m
+    memory: 1010Mi
+  requests:
+    cpu: 300m
```
We should bump this up by default. The initial startup is taking minutes and would likely benefit from more gas.

lodotek replied:
I have upped the values a bit, let's see if that helps..

viglesiasce reviewed (changes requested):
The sapho pod never went to a healthy state. 

The following was found in the log:

```console
Cannot start Sapho:
- Sapho context cannot be started. Missing injection context.
Error creating bean with name 'entityManagerFactory' defined in class path resource [spring-sapho-database.xml]: Cannot resolve reference to bean 'dataSource' while setting bean property 'dataSource'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSource' defined in class path resource [spring-sapho-database.xml]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.zaxxer.hikari.HikariDataSource]: Constructor threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Communications link failure
```

Commit:
Disabled Persistant volume support for mysql due to it causing sapho server to try connecting to mysql before it is ready

lodotek reviewed (commented):
For some reason the mysql container failed to bind the PVC on every first attempt it appeared, resulting in our sapho container from trying to prematurely connect to MySQL. As a temporary solution, I have disable the persistent volume support for MySQL. 

Also tweaked the resource settings...

lodotek commented:
Anything else needed from me to get this PR approved?

prydonius commented:
@lodotek MySQL was failing for me due to `ERROR 1396 (HY000) at line 1: Operation CREATE USER failed for 'root'@'%'` when persistence is enabled. Both MySQL and Sapho started up fine for me when I removed the `mysqlUser: root` setting from `values.yaml` and re-enabled persistence.

If you're having issues with PVCs getting bound, does your cluster support dynamic provisioning?

prydonius commented on incubator/sapho/templates/NOTES.txt:
```diff
@@ -0,0 +1,34 @@
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:{{ .Values.service.externalPort }}
+{{- end }}
+MySQL can be accessed via port 3306 on the following DNS name from within your cluster:
```
I'm not sure you really need to mention how MySQL can be accessed here. If you do though, can you add a line break, as it looks like it's part of the previous statement:

```
  export SERVICE_IP=$(kubectl get svc --namespace default sapho-sapho -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
  echo http://$SERVICE_IP:80
MySQL can be accessed via port 3306 on the following DNS name from within your cluster:
sapho-sapho.default.svc.cluster.local
```

prydonius commented on incubator/sapho/values.yaml:
```diff
@@ -0,0 +1,34 @@
+image: sapho/ops-docker-tomcat:8-jre8
+
+mysql:
+  mysqlUser: root
```
This setting causes MySQL to crash, `root` is the default user created so with this it tries to create the root user twice. It should start up fine without this set.

Commit:
Edited values.yaml:
Removed line 8 - This setting causes MySQL to crash, root is the default user created so
with this it tries to create the root user twice.

NOTES.TXT:
Removed section about MySQL.

prydonius commented on incubator/sapho/templates/deployment.yaml:
```diff
@@ -0,0 +1,49 @@
+        ports:
+        - name: http
+          containerPort: 8080
+#        livenessProbe:
```
@lodotek this can be uncommented again right?

lodotek replied:
Yes, I think you are right. I will test and push the changes.Thanks for all the help, sorry I am kind of a noob haha...

lodotek reviewed (commented):
incubator/sapho/values.yaml:
Removed line 8 specifying mysqlUser
Changed Line 9, re-enabled persistence...

incubator/sapho/templates/NOTES.txt:
Removed section about MySQL

prydonius commented on incubator/sapho/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+description: A Helm chart for Kubernetes
```
Could you change this to a short description of Sapho?

prydonius commented on incubator/sapho/README.md:
```diff
@@ -0,0 +1,95 @@
+# Sapho
+
+[Sapho](https://www.sapho.com/) is awesome.
```
Would be good to have a better description of what Sapho is here.

lodotek replied:
Thank you for the reminder, this has been completed! :-)

prydonius commented on incubator/sapho/README.md:
```diff
@@ -0,0 +1,95 @@
+- Kubernetes 1.4+ with Beta APIs enabled
+- PV provisioner support in the underlying infrastructure
+
+## Get this chart
```
Can you update these sections to use the Helm repo? For example, see: https://github.com/kubernetes/charts/tree/master/stable/mariadb#installing-the-chart

Commit:
Re-enabled livenessProbe (prev disabled for troubleshooting)

prydonius commented:
@lodotek is there any reason you want to keep this in incubator? I would move it to stable since it meets the requirements.

Commit:
incubator/sapho/Chart.yaml:
Added proper chart description and source URLs.

incubator/sapho/README.md:
Updated readme with a product description and revised install instructions.

lodotek commented:
@prydonius Thank you for all you help and suggestions.  Regarding your comment about why am I keeping this in incubator. I thought this PR needs to be approved first, no?

prydonius commented:
@lodotek no worries! As long as a chart meets the requirements listed in [CONTRIBUTING.md](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart) it can be submitted to the stable repo. The incubator repo is for charts that don't meet these requirements (for example, using k8s alpha features), but can be improved on until they are ready to be moved into stable.

So since this already meets the requirements, it can be submitted directly to stable. Sorry for the confusion there.

Commit:
Moved to stable/

Commit:
Moved into stable/

prydonius commented:
@viglesiasce this worked for me, can you check again if you're seeing the same issue you had before?

viglesiasce commented:
Im still seeing the same issue as before when installing in a fresh cluster. This is likely due to the MySQL image being pulled for the first time and taking a while to come up. 

```console
Caused by: java.net.SocketTimeoutException: connect timed out
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_111]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_111]
        at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_111]
        at com.mysql.jdbc.StandardSocketFactory.connect(StandardSocketFactory.java:211) ~[mysql-connector-java-5.1.39.jar:5.1.39]
        at com.mysql.jdbc.MysqlIO.<init>(MysqlIO.java:300) ~[mysql-connector-java-5.1.39.jar:5.1.39]
        ... 63 common frames omitted
00:11:59.735 [localhost-startStop-1] ERROR s.s.SaphoServletContextListener -

********************************************************************************

Cannot start Sapho:
- Sapho context cannot be started. Missing injection context.
Error creating bean with name 'entityManagerFactory' defined in class path resource [spring-sapho-database.xml]: Cannot resolve reference to bean 'dataSource' while setting bean property 'dataSource'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSource' defined in class path resource [spring-sapho-database.xml]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.zaxxer.hikari.HikariDataSource]: Constructor threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
```

Eventually the pod was restarted due to failing its liveness check. At that point it came up fine.

@lodotek can you add something to the NOTES.txt that instructs user to be patient as the app and db come up?

Commit:
Update NOTES.txt

viglesiasce commented:
@k8s-bot e2e test this
